File: Marriot/DatabaseSystems/02-Document-Database-MongoDB/mongodb-mastery-lab/EXAMPLES/python/crud.py
import os
from datetime import datetime, timezone

from pymongo import MongoClient

MONGODB_URI = os.getenv(
    "MONGODB_URI",
    "mongodb://app:app_password@localhost:27017/mongodb_mastery?authSource=mongodb_mastery&replicaSet=rs0&directConnection=true",
)

client = MongoClient(MONGODB_URI, serverSelectionTimeoutMS=5000)
db = client.mongodb_mastery
users = db.users

now = datetime.now(timezone.utc)

users.update_one(
    {"tenantId": "t1", "email": "python-demo@example.com"},
    {
        "$set": {
            "name": "Python Demo",
            "roles": ["USER"],
            "profile": {"city": "Raleigh", "state": "NC"},
            "updatedAt": now,
        },
        "$setOnInsert": {"createdAt": now},
    },
    upsert=True,
)

user = users.find_one(
    {"tenantId": "t1", "email": "python-demo@example.com"},
    {"_id": 0, "email": 1, "name": 1, "profile": 1},
)
print(user)

users.update_one(
    {"tenantId": "t1", "email": "python-demo@example.com"},
    {"$addToSet": {"roles": "EDITOR"}, "$currentDate": {"updatedAt": True}},
)

users.delete_one({"tenantId": "t1", "email": "python-demo@example.com"})
client.close()

File: Marriot/DatabaseSystems/02-Document-Database-MongoDB/mongodb-mastery-lab/EXAMPLES/nodejs/src/crud.js
import { closeDb, getDb } from './db.js';

async function main() {
  const db = await getDb();
  const users = db.collection('users');

  await users.updateOne(
    { tenantId: 't1', email: 'node-demo@example.com' },
    {
      $set: {
        name: 'Node Demo',
        roles: ['USER'],
        profile: { city: 'Denver', state: 'CO' },
        updatedAt: new Date()
      },
      $setOnInsert: { createdAt: new Date() }
    },
    { upsert: true }
  );

  const user = await users.findOne(
    { tenantId: 't1', email: 'node-demo@example.com' },
    { projection: { _id: 0, email: 1, name: 1, profile: 1 } }
  );

  console.log('User:', user);

  await users.updateOne(
    { tenantId: 't1', email: 'node-demo@example.com' },
    { $addToSet: { roles: 'EDITOR' }, $currentDate: { updatedAt: true } }
  );

  await users.deleteOne({ tenantId: 't1', email: 'node-demo@example.com' });
}

main()
  .catch((error) => {
    console.error(error);
    process.exitCode = 1;
  })
  .finally(closeDb);
